netket/utils/group/semigroup.py

```diff
@@ -0,0 +1,169 @@
+# Copyright 2021 The NetKet Authors - All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Ignore false-positives for redefined `product` functions:
+# pylint: disable=function-redefined
+
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+import itertools
+from typing import Callable, List
+
+from plum import dispatch
+
+import numpy as np
+
+from netket.utils import HashableArray
+from netket.utils.types import Array, DType, Shape
+
+
+class ElementBase(ABC):
+    @abstractmethod
+    def __call__(self, arg):
+        pass
+
+    def __matmul__(self, other):
+        return product(self, other)
+
+
+class Element(ElementBase):
+    pass
+
```

Comment:
cc @femtobit who wrote this...

1) Why do we distinguish between ElementBase and Element? I see no reason not to merge the two.
2) I think the distinction between `call` and `matmul` is confusing. I assume that by `call` you intend only to apply the element to an element of the vector space on which the group acts (a ket usually) and matmul is to multiply/combine elements together?

I think we can also conflate the two into the same operation, matmul and use dispatch ( : Array) to distinguish among the two.

Reply:
1. I think this is a left-over from the initial implementation where this was needed to work around some dispatching issue if `Identity` was an `Element` (note how it is only `ElementBase` right now). I think this might not be a problem with the current plum dispatch, tough, so maybe it can be removed now.
2. I don't know. This is conflating two mathematical operations that are independent. I'm pretty sure group theory textbooks at least also like to distinguish this by writing, e.g., `ab` for a product and `a.x` for the action of an element or similar. If having two distinct operations use two method names is confusing, would having to distinct operations use one method name be less confusing? Seems unlikely to me.

Reply:
1. I think it should not be an issue with plum so we can merge the two
2. In physics it's my impression that we represent the action of `T_i` on `x` as `T_i x`. In Cornwell (Group Physics in Physics) the action is also represented by justaxposition. 

The argument of Cornwell is that groups/semigroups are rings/semirings, so they have multiplication (`T_1T_2` follows the axioms of ring multiplciation). Since `T_1(T_2(x)) = (T_1T_2)(x)` due to associativity, then the ordering is irrelevant and the notation can be conflated until one explicitly desires to be more explicit for some reason.

I agree with him that there is no need to distinguish the two operations, one still can by using parenthesis, and dispatch does the rest.

Reply:
re 2. So you suggest dispatching as follows, I assume: If rhs is a subclass of `Element` we delegate to `product`, otherwise to what is currently `__call__`, right?

Reply:
Kinda.
I'd delegate to what is currently `__call__` when hitting a signature `product(el : ElementBase, x : types.Array)`.
`types.Array` is already crafted to catch jax arrays/tracers and numpy arrays so this should work.
If we want it to work with something custom at some point, we/one can add a dispatch rule.
As opposed to it not working at all.